eth-contracts/contracts/CKBChain.sol

```diff
@@ -69,6 +73,33 @@ contract CKBChain is ICKBChain, ICKBSpv {
 
     /// #ICKBSpv
     function proveTxExist(bytes calldata txProofData, uint64 numConfirmations) external view returns (bool){
+        bytes29 proofView = txProofData.ref(uint40(ViewSpv.SpvTypes.CKBTxProof));
+        uint64 blockNumber = proofView.spvBlockNumber();
+        bytes32 blockHash = proofView.blockHash();
+        require(blockNumber + numConfirmations <= lastBlockNumber, "blockNumber is too ahead");
+        require(canonicalHeaderHashes[blockNumber] == blockHash, "blockNumber and blockHash mismatch");
+        require(canonicalTransactionRoots[blockHash] != bytes32(0), "blockHash invalid or too old");
+        uint16 index = proofView.txMerkleIndex();
+        uint16 sibling;
+        uint256 lemmasIndex = 0;
+        bytes29 lemmas = proofView.lemmas();
+        uint256 length = lemmas.len() / 32;
+        bytes32 res = proofView.txHash();
+
+        while (lemmasIndex < length && index > 0) {
+            sibling = ((index + 1) ^ 1) - 1;
+            if (index < sibling) {
+                res = CKBCrypto.digest(abi.encodePacked(res, lemmas.indexH256Array(lemmasIndex), new bytes(64)), 64);
+            } else {
+                res = CKBCrypto.digest(abi.encodePacked(lemmas.indexH256Array(lemmasIndex), res, new bytes(64)), 64);
+            }
+
+            lemmasIndex++;
+            // index = parent(index)
+            index = (index - 1) >> 1;
+        }
+        res = CKBCrypto.digest(abi.encodePacked(res, proofView.witnessesRoot(), new bytes(64)), 64);
+        require(res == canonicalTransactionRoots[res], "proof not passed");
```

Comment:
Does this pass the test? Seems like a bug.
There are 2 `res` here,  `proofView.txHash()` and the one calculated from proof.
Should use better var name.

Reply:
this `res` is used to record the result of transactionsRoot, `res` is always the intermediate variable during the whole calculation

Reply:
Rename the 2 `res` to make them more specific, you can use `raw_transactions_root` and `transactions_root`, ref by <https://github.com/nervosnetwork/ckb/blob/582cd7cdd2cef3320c83382b83845e14c37d7529/util/types/src/extension/shortcuts.rs#L153-L155>.

There are 2 rules here:
1. Do not use var name `res`, make it more specific.
2. Do not reuse the var `res` here. It is not like reuse `sum` in a loop, they are totally different here, just happened to be the same type.

Reply:
fixed

Reply:
add  // TODO optimize rawTxRoot calculation with assembly code for proveTxExist